pkg/cloudmeta/aws.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (C) 2024 ScyllaDB
+
+package cloudmeta
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/pkg/errors"
+)
+
+// AWSMetadata is a wrapper around ec2 metadata client.
+type AWSMetadata struct {
+	ec2meta *ec2metadata.EC2Metadata
+}
+
+// NewAWSMetadata is a constructor for  AWSMetadata service.
+func NewAWSMetadata() (*AWSMetadata, error) {
+	session, err := session.NewSession()
+	if err != nil {
+		return nil, errors.Wrap(err, "session.NewSession")
+	}
+	return &AWSMetadata{
+		ec2meta: ec2metadata.New(session),
+	}, nil
+}
+
+// NewAWSMetadataWithEndpoint is a constructor for  AWSMetadata service that allows your to overwrite default metadata endpoint.
+// Might be useful for tests.
+func NewAWSMetadataWithEndpoint(endpoint string) (*AWSMetadata, error) {
+	session, err := session.NewSession()
+	if err != nil {
+		return nil, errors.Wrap(err, "session.NewSession")
+	}
+	cfg := aws.NewConfig()
+	if endpoint != "" {
+		cfg = cfg.WithEndpoint(endpoint)
+	}
+	return &AWSMetadata{
+		ec2meta: ec2metadata.New(session, cfg),
+	}, nil
+}
+
+// Metadata return InstanceMetadata from aws if available.
+func (aws *AWSMetadata) Metadata(ctx context.Context) (InstanceMetadata, error) {
+	if !aws.ec2meta.AvailableWithContext(ctx) {
+		return InstanceMetadata{}, errors.New("metadata is not available")
```

Comment:
Can you extract `errors.New("metadata is not available")` to exported var ErrMetadataNotAvailable ?

Reply:
You need to call errors.WithStackTrace(ErrMetadataNotAvailable) later on, when returning this error.

Reply:
This is expected that the metadata might not be available, as we don't know which provider is used.
Adding stack to this kind of error will probably just clog the logs. Why do you think that the stack is useful here?

Reply:
That's valid statement. We are probing all providers. OK, not needed here.

Reply:
Or actually, we know it form the `--location` provider part.
So there is no need to query all of the providers, which means that there is no need for the `CloudMeta` at all?

Reply:
But theoretically, you can have VMs deployed whenever you want and use the other location than the cloud provider you use for your VMs.
It's more solid to just probe it instead of relying on `--location`.